adds doc providing additional advice to contributors

adammichaelwood opened the pull request:
new doc on tips for contributors

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,142 @@
+Smallest meaningful PR
+------------------------
+
+A PR should normally address one issue. This makes it easier to review, easier to deploy, and easier to roll back in case of a problem. Additionally, the smaller the PR, the less likely it is to create a merge issue.
```
This makes it easier to review, deploy and roll back in case of an **error**. -- error seems a better word while deployment and roll backs are considered.

adammichaelwood replied:
I disagree because "error" is a specific condition created by build process. We might roll back because something was factually wrong. Or because we want to organize something differently. Or lot so of other *problems*.

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,142 @@
+
+A PR should normally address one issue. This makes it easier to review, easier to deploy, and easier to roll back in case of a problem. Additionally, the smaller the PR, the less likely it is to create a merge issue.
+
+The exception is when several issues are closely related or can reasonably be worked on together.  In this case, it should be clear by looking at the conversation on the Github issues that the items are related and will be worked on together.
```
There can be a point about several issues linked together. Like *Best Practices/Tips issue* and *Adding Android level security tips* are interconnected. It's possible a PR can address one issue completely and partly resolve other. How can that be addressed while describing the PR?

adammichaelwood replied:
I'll add a sentence about making it clear in the PR as well as the issues.

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,142 @@
+When working locally, commit often. Don't wait until you have 100 lines of changes across multiple files.
+
+- If you need to copy or move a large section or file, commit that change before also editing it. 
+- If you have to create a new template file based on an existing template file, copy the file in one commit and then work on the changes. This makes it easier to know what you actually did.
```
If you have to create a new template file based on an existing one ( just trying to minimize words here )

adammichaelwood replied:
i'll think about it

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,142 @@
+
+  - Asking questions is always better than guessing.
+
+- Submit a ``[WIP]`` (work in progress) pull request.
```
Mentioning these PR titles should have *[WIP]* written before the heading should clear this more

adammichaelwood replied:
yes, i need to make that more clear somewhere

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,142 @@
+It is okay to abandon an issue
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Sometimes you simply cannot complete work you have said you were going to complete. This could happen because you don't have the skills or knowledge to complete the work, or because the issue cannot actually be completed as scoped, or because you don't have the time.
```
don't have all the required skills or knowledge sounds better :P

adammichaelwood replied:
> because you don't have all the required skills or knowledge to complete the work

or

> because you don't have the required skills or knowledge,

?

And why do you think so?

mish24 replied:
A person might have few of the required skills but maybe not related to a specific topic like Android (considering this is documentation)

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,142 @@
+
+All PRs must be directly connected to open issues. PRs should not represent suggestions, good ideas, or independent intiative. 
+
+If you have a good idea, file an issue. If you are curious about whether something should be an issue, chat with one of the core team in the `#docs-code` channel on the `Slack <https://opendatakit.slack.com>`_.
```
Adding *wait for a collaborator or admin to comment on the new issue before you start working on it* can also help. Since people open issues themselves and submit a PR before the issue is validated

adammichaelwood replied:
This is a very good point.

mish24 replied:
I think mentioning the forum is also important. Involved in the forum kind of keeps you in touch with what's actually going on in other projects (new release of Collect, updates on Outreachy etc)

adammichaelwood commented:
@mish I've pushed a commit that addresses your comments. (Well... the ones I agreed with, anyway.)

I also added another tip.

If you have time for another review round, I'd appreciate it.

troublemagnet commented on contributing-tips.rst:
```diff
@@ -0,0 +1,174 @@
+
+A PR title should answer the question, "What does this Pull Request do?"
+
+Good PR titles:
```
I guess a good PR title should also have a character limit.

troublemagnet commented on contributing-tips.rst:
```diff
@@ -0,0 +1,174 @@
+
+A PR should normally address one issue. This makes it easier to review, easier to deploy, and easier to roll back in case of a problem. Additionally, the smaller the PR, the less likely it is to create a merge issue.
+
+The exception is when several issues are closely related or can reasonably be worked on together.  In this case, it should be clear by looking at the conversation on the Github issues that the items are related and will be worked on together. Your PR message should also make it clear which issues are being worked on, and whether the PR closes the issues or not.
```
Adding an example of issue reference like "Closes https://github.com/opendatakit/docs/issues/{no}" will be a good idea I guess. 
We can frame it this way :
Each PR description must end with "Closes https://github.com/opendatakit/docs/issues/{no}" if it closes the issue or "Addresses https://github.com/opendatakit/docs/issues/{no}" if there's work remaining.

troublemagnet replied:
Also I don't know um maybe we could have a standard format for a pull request? For example: 
PR title : It should explain what the PR is addressing and not how, and why. The character limit could be decided.
PR description : A short 4-5 lines paragraph briefly explaining how it has been done.
Issue reference : Closes/Addresses/Fixes {url} 
where url is the complete url of issue being worked on.

troublemagnet commented on contributing-tips.rst:
```diff
@@ -0,0 +1,174 @@
+
+If you want a review, comment, opinion, or help on a ``[WIP]`` PR, please say so.
+
+If you finish the work and want the PR to be merged, you do not need to open a new one. Just edit the PR title. 
```
Um I guess we can replace the "please say so" by "tag the concerned person"?

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,174 @@
+- Submit a ``[WIP]`` (work in progress) pull request.
+
+  - If we can see what progress you have made, it is easier to offer help.
+  - Even if you don't complete the project, perhaps someone else can pull in your in-progress work and build on it.
```
don't complete the *task*

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,174 @@
+
+Sometimes you simply cannot complete work you have said you were going to complete. This could happen because you don't have all the required skills or knowledge to complete the work, or because the issue cannot actually be completed as scoped, or because you don't have the time.
+
+Please let the community know in the issue discussion. This way, everyone knows that someone else can take up the project (or that we need to rethink it).
```
in the issue discussion, and once in the related PR as well? Since most of the e-mails we get are comment threads in a PR.

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,174 @@
+
+.. _discuss-issues:
+
+Discuss issues before working
```
Should this include the **bot** thing? In my personal opinion we should completely drop that, else people can just do `opendatakitbot-claim` before discussing. But if bot is here to stay, it is better to mention that step here.

mish24 commented on contributing-tips.rst:
```diff
@@ -0,0 +1,174 @@
+
+You can contribute to this community, no matter your background or skills. 
+
+- If there is something you don't know how to do, you can ask.  
```
ask either on the issue, slack channel or the forum
Forum can also be mentioned as a source to clear doubts and get different ideas.

mish24 commented:
@adammichaelwood I think that with slack's link we can give the link to forum also. It is not mentioned anywhere in docs. It's really helpful in understanding how ODK people have been working all this time.

adammichaelwood commented:
@yanokwa I think this is ready to merge

mish24 commented:
@adammichaelwood Just one doubt here. When do we need to provide screenshots? It's mentioned that image must be cropped showing all the relevant info, but in what cases it's definitely needed and when is it not?

adammichaelwood commented:
So far I haven't found any need to include screen shots for things. I imagine when we get to documenting Aggregate, that will become necessary. But I can't think of what specific advice to give for that until we get to it.

yanokwa commented on contributing-tips.rst:
```diff
@@ -0,0 +1,199 @@
+Share Work in Progress
+-------------------------
+
+It can be helpful to share your in-progress work. To mark a PR as a work in progress, append ``[WIP]`` to the beginning of the PR title. We will not merge ``[WIP]`` PRs, and we won't do a review on them unless you ask. 
```
Why make people type the brackets? Just `WIP:` should be enough, no?

adammichaelwood replied:
> Why...?

Just what I've seen on other repos. Doesn't really matter

> should be enough...

sure.

yanokwa commented on contributing-tips.rst:
```diff
@@ -0,0 +1,199 @@
+
+- If there is something you don't know how to do, you can ask. 
+  - If it is issue related, ask on the issue.
+  - If it is more general, try the #docs-code channel in the `ODK Developer Slack <https://opendatakit.slack.com>`.
```
This does not show up as a link. Also we should use `http://slack.opendatakit.org` (no HTTPS). That link lets new contributors invite themselves to the slack.

adammichaelwood replied:
that explains so many questions I've gotten about invites...

yanokwa commented on contributing-tips.rst:
```diff
@@ -0,0 +1,199 @@
+
+If you decide to work on an issue, let the community know you are working on it by *claiming* the issue.
+
+> @opendatakit-bot claim
```
In my local build, this doesn't show up as a quote. Should it?

adammichaelwood replied:
Oh my fault- that's the Markdown syntax for blockquotes. 
rst block quotes are just indented

yanokwa commented on contributing-tips.rst:
```diff
@@ -0,0 +1,199 @@
+
+You can contribute to this community, no matter your background or skills. 
+
+- If there is something you don't know how to do, you can ask. 
```
These don't show up in my local build as indented bullets.

adammichaelwood replied:
syntax typo -- needs a line break

yanokwa commented on contributing-tips.rst:
```diff
@@ -0,0 +1,199 @@
+- If there is something you don't know how to do, you can ask. 
+  - If it is issue related, ask on the issue.
+  - If it is more general, try the #docs-code channel in the `ODK Developer Slack <https://opendatakit.slack.com>`.
+  - You can also send a private message to @adammichaelwood or another maintainer. 
```
I think this should be cut. We should default to communicating out in the open because it reduces the burden on maintainers in the long term. That is, everyone can see the question, everyone can have a chance to answer, and the answer will be seen by everyone.

adammichaelwood replied:
Makes sense. My thinking was to try to be welcoming to people who might feel embarrassed or something, but probably they can figure out how/who to PM if they really want to.

yanokwa commented on contributing-tips.rst:
```diff
@@ -0,0 +1,199 @@
+
+.. _small-commits:
+
+Small Atomic Commits
```
For consistency, with the rest of the titles, his should probably be title case

adammichaelwood replied:
I am so bad at following casing rules.

(Need to add this to style guide and also lint for this.)

adammichaelwood commented:
@yanokwa Review notes addressed, plus a couple other things.
Please re-review and merge at your convenience.
